Sources/WhisperKit/MLX/MLXFeatureExtractor.swift

```diff
@@ -0,0 +1,51 @@
+//  For licensing see accompanying LICENSE.md file.
+//  Copyright © 2024 Argmax, Inc. All rights reserved.
+
+import Foundation
+import MLX
+import CoreML
+import WhisperKit
+
+@available(macOS 13, iOS 16, watchOS 10, visionOS 1, *)
+open class MLXFeatureExtractor: FeatureExtracting {
+    public let melCount: Int?
+    private let nFFT: Int
+    private let hopLength: Int
+    private let filters: MLXArray
+
+    public init(
+        melCount: Int = 80,
+        nFFT: Int = 400,
+        hopLength: Int = 160
+    ) {
+        self.melCount = melCount
+        self.nFFT = nFFT
+        self.hopLength = hopLength
+        self.filters = loadMelFilters(nMels: melCount)
+    }
+
+    public func logMelSpectrogram(fromAudio inputAudio: MLMultiArray) async throws -> MLMultiArray? {
+        try Task.checkCancellation()
+        let input = inputAudio.withUnsafeBytes { ptr in
+            MLXArray(ptr, inputAudio.shape.map { $0.intValue }, type: Float.self)
+        }
+        let logMelSpectrogram = WhisperKitMLX.logMelSpectrogram(
+            audio: input,
+            filters: filters,
+            nMels: melCount ?? 80,
+            nFFT: nFFT,
+            hopLength: hopLength
+        )
+        let logMelSpectrogramData = logMelSpectrogram.asArray(Float.self)
+        let result = try MLMultiArray(shape: logMelSpectrogram.shape.map { NSNumber(value: $0) }, dataType: .float32)
+        let maxRow = logMelSpectrogram.shape[0]
+        let maxCol = logMelSpectrogram.shape[1]
+        for row in 0..<maxRow {
+            for col in 0..<maxCol {
+                let key = [row, col] as [NSNumber]
+                result[key] = NSNumber(value: logMelSpectrogramData[row * maxCol + col])
+            }
+        }
+        return result
```

Comment:
maybe there is a faster way to create a `MLMultiArray` from `MLXArray`? we could ask the `MLX` team about it

Reply:
@davidkoski @awni Curious to hear your thoughts on this if you have any

Reply:
Check out [asData](https://swiftpackageindex.com/ml-explore/mlx-swift/main/documentation/mlx/mlxarray/asdata(nocopy:)), that might be useful here especially if you can avoid the copy.

Reply:
came up with something like this

```swift
extension MLXArray {
    func asMLMultiArray(
        dataType: MLMultiArrayDataType = .float32,
        noCopy: Bool = true
    ) throws -> MLMultiArray {
        var data = asData(noCopy: noCopy)
        return try data.withUnsafeMutableBytes { (ptr: UnsafeMutableRawBufferPointer) in
            try MLMultiArray(
                dataPointer: ptr.baseAddress!,
                shape: shape.map { NSNumber(value: $0) },
                dataType: dataType,
                strides: strides.map { NSNumber(value: $0) }
            )
        }
    }
}
```

Reply:
I wasn't sure about using Data as it is a struct and you are meant to use the pointer inside the callback.  I think something like this would probably work (it compiles but not tested):

```swift
        let array = MLXArray.ones([3, 4, 5])
        
        let coreMLType: MLMultiArrayDataType
        switch array.dtype {
        case .bool, .bfloat16, .complex64: fatalError("Unsupported type: \(array.dtype)")
        case .uint8, .uint16, .uint32, .uint64: fatalError("Unsupported type: \(array.dtype)")
        case .int8, .int16, .int64: fatalError("Unsupported type: \(array.dtype)")
        case .int32: coreMLType = .int32
        case .float16: coreMLType = .float16
        case .float32: coreMLType = .float32
        }
        
        // a buffer to be passed to CoreML
        let buffer = UnsafeMutableRawPointer.allocate(byteCount: array.nbytes, alignment: 8)
        
        // copy the data from the MLXArray backing into buffer
        array.asData(noCopy: true).withUnsafeBytes { ptr in
            let destination = UnsafeMutableRawBufferPointer(start: buffer, count: array.nbytes)
            ptr.copyBytes(to: destination)
        }
        
        func free(buffer: UnsafeMutableRawPointer) {
            buffer.deallocate()
        }
        
        let coreMLArray = try MLMultiArray(dataPointer: buffer, shape: array.shape.map { NSNumber(value: $0) }, dataType: coreMLType, strides: array.strides.map { NSNumber(value: $0) }, deallocator: free(buffer:))
```

Reply:
will look into it, thanks a lot @davidkoski